File: src/chain-observers/collectors/common/operation-collector.ts
import { BlockClassifier, OperationClassifier } from "../../classifiers";
import { TRegisterEvaluationContext } from "../../classifiers/collector-classifier-base";
import { IOperationTransactionPair } from "../../classifiers/operation-classifier";
import { TCollectorEvaluationContext } from "../../factories/data-evaluation-context";
import { CollectorBase, TAvailableClassifiers } from "../collector-base";

export class OperationCollector extends CollectorBase<OperationClassifier> {
  public usedContexts(): Array<TRegisterEvaluationContext> {
    return [BlockClassifier];
  }

  public async get(data: TCollectorEvaluationContext) {
    const { transactions } = await data.get(BlockClassifier);

    const operations: Array<IOperationTransactionPair> = [];
    const operationsPerType: Record<string, Array<IOperationTransactionPair>> = {};

    const startOperationPerType = Date.now();

    for(const transaction of transactions)
      for(const operation of transaction.transaction.operations) {
        operations.push({
          operation,
          transaction
        });

        let opType = "";
        for(const type in operation)
          if(operation[type] !== undefined) {
            opType = type;
            break;
          }

        if(operationsPerType[opType] === undefined)
          operationsPerType[opType] = [];

        operationsPerType[opType].push({
          operation: operation[opType],
          transaction
        });
      }

    data.addTiming("operationPerType", Date.now() - startOperationPerType);

    return {
      /*
       * Instruct TypeScript typings that OperationClassifier.name is actualy a Classifier name we expect.
       * This is required for the bundlers to properly deduce the type of the classifier in data evaluation context.
       */
      [OperationClassifier.name as "OperationClassifier"]: {
        operations,
        operationsPerType
      } as TAvailableClassifiers["OperationClassifier"]
    };
  };
}

File: python/workerbee/chain_observers/collectors/common/operation_collector.py
"""OperationCollector — extracts operations from block transactions.

Mirrors src/chain-observers/collectors/common/operation-collector.ts.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from ...classifiers.block_classifier import BlockClassifier
from ...classifiers.collector_classifier_base import TRegisterEvaluationContext
from ...classifiers.operation_classifier import OperationClassifier
from ...payloads import OperationsPerType, OperationTransactionPair
from ..collector_base import CollectorBase

if TYPE_CHECKING:
    from ...factories.data_evaluation_context import TCollectorEvaluationContext


class OperationCollector(CollectorBase):
    def used_contexts(self) -> list[TRegisterEvaluationContext]:
        return [BlockClassifier]

    async def get(self, data: TCollectorEvaluationContext) -> dict[str, Any]:
        block_data = await data.get(BlockClassifier)
        transactions = block_data["transactions"]

        operations: list[OperationTransactionPair] = []
        operations_per_type: OperationsPerType = {}

        with data.add_timing("operation_per_type"):
            for tx_data in transactions:
                tx = tx_data["transaction"]
                tx_ops = tx.get("operations", []) if isinstance(tx, dict) else getattr(tx, "operations", [])
                for operation in tx_ops:
                    # WorkerBee uses only hf26 APIs (block_api/database_api), so each
                    # operation is the hf26 envelope `Operation` (op.type / op.value).
                    # The flat `operations` list keeps the envelope; `operations_per_type`
                    # keys by op.type and stores the unwrapped op.value body.
                    operations.append({"operation": operation, "transaction": tx_data})
                    operations_per_type.setdefault(operation.type, []).append(
                        {
                            "operation": operation.value,
                            "transaction": tx_data,
                        }
                    )

        return {
            OperationClassifier.__name__: {
                "operations": operations,
                "operations_per_type": operations_per_type,
            },
        }
